Generation fails when Factory method is in base interface and uses generics

```kotlin
interface BaseThing<T> {
  interface Factory {
    fun create(view: View): BaseThing<T>
  }
}

class StringThing @AssistedInject constructor (
  @Assited view: View
): BaseThing<String> {

  @AssistedInject.Factory
  interface Factory : BaseThing.Factory<String>
}
```

This will not generate an implementation for the factory, but a generated Dagger 2 module will still try to reference the not generated factory.
Adding `override fun create(view: View): BaseThing<String>` to the Factory will fix it.

Yeah I'm getting 
```
/Users/gabriel/.../SearchBinder_AssistedFactory.java:26: error: cannot find symbol
  public BaseBinder<S, E> create(View view) {
                    ^
  symbol:   class S
  location: class SearchBinder_AssistedFactory

/Users/gabriel/....w/SearchBinder_AssistedFactory.java:26: error: cannot find symbol
  public BaseBinder<S, E> create(View view) {
                       ^
  symbol:   class E
  location: class SearchBinder_AssistedFactory
```

--- a/assisted-inject-processor/src/main/java/com/squareup/inject/assisted/processor/AssistedInjectProcessor.kt
+++ b/assisted-inject-processor/src/main/java/com/squareup/inject/assisted/processor/AssistedInjectProcessor.kt
@@ -43,6 +43,8 @@ import javax.lang.model.element.ExecutableElement
 import javax.lang.model.element.Modifier.PRIVATE
 import javax.lang.model.element.Modifier.STATIC
 import javax.lang.model.element.TypeElement
+import javax.lang.model.type.DeclaredType
+import javax.lang.model.type.ExecutableType
 import javax.lang.model.util.Elements
 import javax.lang.model.util.Types
 import javax.tools.Diagnostic.Kind.ERROR
@@ -231,8 +233,14 @@ class AssistedInjectProcessor : AbstractProcessor() {
       }
     }
 
+    // Project the factory method (which may have come from a supertype) as if it were a member of
+    // the subtype to resolve any generic parameters.
+    val factoryExecutable = types.asMemberOf(factoryType.asType() as DeclaredType,
+        factoryMethod) as ExecutableType
+
     val expectedKeys = assistedRequests.map { it.namedKey }.sorted()
-    val factoryKeys = factoryMethod.parameters.map { it.asNamedKey() }
+    val factoryKeys = factoryMethod.parameters
+        .zip(factoryExecutable.parameterTypes) { element, mirror -> element.asNamedKey(mirror) }
     val keys = factoryKeys.sorted()
     if (keys != expectedKeys) {
       val message = buildString {
@@ -255,7 +263,7 @@ class AssistedInjectProcessor : AbstractProcessor() {
 
     val targetType = targetType.asType().toTypeName()
     val factoryType = factoryType.toClassName()
-    val returnType = factoryMethod.returnType.toTypeName()
+    val returnType = factoryExecutable.returnType.toTypeName()
     val methodName = factoryMethod.simpleName.toString()
     val generatedAnnotation = createGeneratedAnnotation(elements)
     return AssistedInjection(targetType, requests, factoryType, methodName, returnType,

--- a/assisted-inject-processor/src/main/java/com/squareup/inject/assisted/processor/Key.kt
+++ b/assisted-inject-processor/src/main/java/com/squareup/inject/assisted/processor/Key.kt
@@ -20,6 +20,7 @@ import com.squareup.inject.assisted.processor.internal.toTypeName
 import com.squareup.javapoet.AnnotationSpec
 import com.squareup.javapoet.TypeName
 import javax.lang.model.element.VariableElement
+import javax.lang.model.type.TypeMirror
 
 private val keyComparator = compareBy<Key>({ it.type.toString() }, { it.qualifier == null })
 
@@ -33,7 +34,7 @@ data class Key(
 }
 
 /** Create a [Key] from this type and any qualifier annotation. */
-fun VariableElement.asKey() = Key(asType().toTypeName(),
+fun VariableElement.asKey(mirror: TypeMirror = asType()) = Key(mirror.toTypeName(),
     annotationMirrors.find {
       it.annotationType.asElement().hasAnnotation("javax.inject.Qualifier")
     }?.let { AnnotationSpec.get(it) })

--- a/assisted-inject-processor/src/main/java/com/squareup/inject/assisted/processor/NamedKey.kt
+++ b/assisted-inject-processor/src/main/java/com/squareup/inject/assisted/processor/NamedKey.kt
@@ -1,6 +1,7 @@
 package com.squareup.inject.assisted.processor
 
 import javax.lang.model.element.VariableElement
+import javax.lang.model.type.TypeMirror
 
 private val namedKeyComparator = compareBy<NamedKey>({ it.key }, { it.name })
 
@@ -14,4 +15,4 @@ data class NamedKey(
 }
 
 /** Create a [NamedKey] from this type, any qualifier annotation, and the name. */
-fun VariableElement.asNamedKey() = NamedKey(asKey(), simpleName.toString())
+fun VariableElement.asNamedKey(mirror: TypeMirror = asType()) = NamedKey(asKey(mirror), simpleName.toString())

--- a/assisted-inject-processor/src/test/java/com/squareup/inject/assisted/processor/AssistedInjectProcessorTest.kt
+++ b/assisted-inject-processor/src/test/java/com/squareup/inject/assisted/processor/AssistedInjectProcessorTest.kt
@@ -1331,7 +1331,6 @@ class AssistedInjectProcessorTest {
 
       import java.lang.Long;
       import java.lang.Override;
-
       import java.lang.String;
       import $GENERATED_TYPE;
       import javax.inject.Inject;
@@ -1387,7 +1386,6 @@ class AssistedInjectProcessorTest {
 
       import java.lang.Long;
       import java.lang.Override;
-
       import java.lang.String;
       import $GENERATED_TYPE;
       import javax.inject.Inject;
@@ -1402,6 +1400,61 @@ class AssistedInjectProcessorTest {
         }
 
         @Override public BaseThing create(String bar) {
+          return new Test(foo.get(), bar);
+        }
+      }
+    """)
+
+    assertAbout(javaSource())
+        .that(input)
+        .processedWith(AssistedInjectProcessor())
+        .compilesWithoutError()
+        .and()
+        .generatesSources(expected)
+  }
+
+  @Test fun factoryMethodInSupertypeWithGenerics() {
+    val input = JavaFileObjects.forSourceString("test.Test", """
+      package test;
+
+      import com.squareup.inject.assisted.Assisted;
+      import com.squareup.inject.assisted.AssistedInject;
+
+      interface BaseThing<T> {
+        interface Factory<T> {
+          BaseThing<T> create(T bar);
+        }
+      }
+
+      class Test implements BaseThing<String> {
+        @AssistedInject
+        Test(Long foo, @Assisted String bar) {}
+
+        @AssistedInject.Factory
+        interface Factory extends BaseThing.Factory<String> {
+        }
+      }
+    """)
+
+    val expected = JavaFileObjects.forSourceString("test.Test_AssistedFactory", """
+      package test;
+
+      import java.lang.Long;
+      import java.lang.Override;
+      import java.lang.String;
+      import $GENERATED_TYPE;
+      import javax.inject.Inject;
+      import javax.inject.Provider;
+
+      $GENERATED_ANNOTATION
+      public final class Test_AssistedFactory implements Test.Factory {
+        private final Provider<Long> foo;
+
+        @Inject public Test_AssistedFactory(Provider<Long> foo) {
+          this.foo = foo;
+        }
+
+        @Override public BaseThing<String> create(String bar) {
           return new Test(foo.get(), bar);
         }
       }

